File: Algorithm/Algorithm_JAVA/src/Shell.java
public class Shell {
    public void shellSort(int[] data) {
        int size = data.length;
        int temp;
        int gap = size;

        for(; gap > 1;) {
            gap = (gap / 3) + 1;
            System.out.println("gap: " + gap);
            for(int i = 0; i < gap; i++) {
                // 삽입 정렬
                for(int j = i + gap; j < size; j+= gap) {
                    // i : gap으로 나눈 묶음을 구분
                    // j : 삽입 정렬 종료 인덱스
                    for(int x = i; x < j; x += gap) {
                        if(data[x] > data[j]) {
                            temp = data[x];
                            data[x] = data[j];
                            data[j] = temp;
                        }
                    }
                }
            }
        }

        for(int i=0; i < size; i++) {
            System.out.print(data[i] + " ");
        }
    }
}

File: Algorithm/Algorithm_C++/Algorithm/shell.cpp
#include <iostream>

using namespace std;

void shellSort(int arr[], int len) {
	int i, j, temp;
	int gap = len / 2;
	while (gap > 0) {
		for (i = gap; i < len; i++) {
			temp = arr[i];
			j = i;
			while (j >= gap && arr[j - gap] > temp) {
				arr[j] = arr[j - gap];
				j -= gap;
			}
			arr[j] = temp;
		}
		gap /= 2;
	}
}

int main() {
	int arr[] = { 10,8,6,20,4,3,22,1,0,15,16 };
	int len = sizeof(arr) / sizeof(int);

	cout << "정렬 전: ";
	for (int x : arr)
		cout << x << " ";
	cout << endl;

	shellSort(arr, len);

	cout << "정렬 후: ";
	for (int x : arr)
		cout << x << " ";
	cout << endl;
}